### DISCUSSION TITLE
Confused about vuetify layout in drawer

### DISCUSSION THREAD

**Discussion by fdinhSG** [Kitware/trame#242] [Q&A] [answered]
I'm trying to build a web app using trame with data viz using pyvista.

What I'm expecting to get:

![What I'm expecting to get](https://user-images.githubusercontent.com/128691970/233589618-2646db9d-7129-4fb0-a3e3-284a2c19fd80.png)

What I'm actually getting:

![What I'm actually getting](https://user-images.githubusercontent.com/128691970/233589921-6dc1360d-d8e4-4a8d-ad80-db2fd5a8cf6a.png)

And this is my code:

```python
from trame.app import get_server
from trame.ui.vuetify import SinglePageWithDrawerLayout
from trame.widgets import vuetify

import pyvista as pv
from pyvista.trame.ui import plotter_ui

# -----------------------------------------------------------------------------
# Trame initialization
# -----------------------------------------------------------------------------

pv.OFF_SCREEN = True

server = get_server()
state, ctrl = server.state, server.controller

state.trame__title = "My App"
ctrl.on_server_ready.add(ctrl.view_update)


# -----------------------------------------------------------------------------
# Plotting
# -----------------------------------------------------------------------------

pl = pv.Plotter()
mesh_actor = None


def create_mesh(solid):
    global pl, mesh_actor
    if mesh_actor is not None:
        pl.remove_actor(mesh_actor)
    if solid == 'sphere':
        mesh_actor = pl.add_mesh(pv.Sphere())
    elif solid == 'cube':
        mesh_actor = pl.add_mesh(pv.Cube())
    pl.reset_camera()


# -----------------------------------------------------------------------------
# GUI
# -----------------------------------------------------------------------------


with SinglePageWithDrawerLayout(server) as layout:
    layout.title.set_text(state.trame__title)

    with layout.drawer as drawer:
        with vuetify.VRow():
            btn_save = vuetify.VBtn('SAVE', click=lambda: print('save'), disabled=True)
            btn_load = vuetify.VBtn('LOAD', click=lambda: print('load'))
            btn_solve = vuetify.VBtn('SOLVE', click=lambda: print('solve'), disabled=True)
            btn_snack = vuetify.VBtn("SNACK", click=lambda: vuetify.VSnackbar('hi'))

        with vuetify.VRow():
            vuetify.VSelect(
                label="Choose 3",
                v_model=("Choose3", 0),
                items=(
                    "choose3",
                    [
                        {"text": "choice 1", "value": 0},
                        {"text": "choice 2", "value": 1},
                        {"text": "choice 3", "value": 2},
                    ],
                ),
            )
            vuetify.VSelect(
                label="Choose 2",
                v_model=("Choose2", 0),
                items=(
                    "choose2",
                    [
                        {"text": "choice 1", "value": 0},
                        {"text": "choice 2", "value": 1},
                    ],
                ),
            )
        
        update_bounds = lambda: print(f'Pmin={state.Pmin}, Pmax={state.Pmax}')
        with vuetify.VRow():
            vuetify.VTextField(label='Text', v_model=('Test','test'))
            vuetify.VTextField(label='Min P', v_model=('Pmin', 0), type='number', single_line=True, change=update_bounds) 
            vuetify.VTextField(label='Max P', v_model=('Pmax', 0), type='number', single_line=True, change=update_bounds) 

    with layout.toolbar:
        vuetify.VSpacer()

        vuetify.VBtn("Sphere", click=lambda: create_mesh("sphere"))

        vuetify.VBtn("Cube", click=lambda: create_mesh("cube"))

    with layout.content:
        with vuetify.VContainer(
            fluid=True,
            classes="pa-0 fill-height",
        ):
            # Use PyVista UI template for Plotters
            view = plotter_ui(pl)
            ctrl.view_update = view.update

server.start()
```

I'm mostly confused about what's going on inside the drawer, in lines 48-84. Specifically:
- Why are all the elements seemingly outside of the borders of the drawer?
- Why aren't the two VSelect stacking horizontally inside the VRow?
- Same for the three VTextField below
- Why is the label applying for the textual VTextField, but not for the two numeric ones?
- How can I make the VSnackbar work? I'd like a popup notification to come out, as explained in [this section](https://vuetifyjs.com/en/components/snackbars/), but I'm not sure I'm getting the grammar right. The only thing I'm getting is a `JS Error => [object Object]`

Thank you!

**Comment by jourdain** [comment, marked as answer]
Welcome to trame!

- First of all Container/Row/Col are supposed to be used together to properly handle their negative margin. 
- You can read more about their grid system [here](https://v2.vuetifyjs.com/en/components/grids/).
- If you want to play and figure out the layout you can use [Play Vuetify](https://play.vuetifyjs.com/). Technically, you just need to nest your widgets in VCol like in that [full example](https://play.vuetifyjs.com/#eNqtU01vwjAM/StVzrTRvjQJMcTumzSJ3QiHtDVaRppGSeiYEP99TlpGWyZUDdpL/WK/Z7/Gix151jqpNkDGZOKg0JI7mDIVRZMq5lqHzxAUXCiMmPMvxlkpf8MAmPKrBbRyPMHhQTB1Kkplma2jXFieSsinc17BhIajTr7HujoDaKcvJc+vxtZqspTX7HKueLYeRueRrrvn7D5CAbQgIXMR7Safa/l/BBc16WDr4pUAmUeSpyCfGHlHiJEL+v6D81Wo6A1Jr8zKt8NY+xYdMw7/vVmyOqjXb0Jba0lGRBS6NA4zdfJpS4V7u/MFrDmwjIyjgHgMF9vHjHw4p+2Y0ixXWJaDFJVJFDiqdEFnmEbNRjlRQJyXxewuuU3uHylefdfGE7BFnOIEFgyyMDJq6Tix+h6g1WSixE1y81BLNFiMhlqvccJNUbACExtQORgwQ2fqlXXm6p2dzObl90zt0XIp1Nr23M5scHpx6PKyoQObp1p6xf2I1J2idl1Clj9tqei8).

Then for the last question, you can not create a detached widget (Snackbar) as a callback. The UI should be created once and rely on the shared state to show/hide/enable UI element. 
In short you are looking at doing something like.

```python
with layout:
   vuetify.VBtn("ToggleSnack", click="show_snack = !show_snack")
   vuetify.VSnackbar("Hello world", v_model=("show_snack", False))
``` 

That part is documented [here](https://v2.vuetifyjs.com/en/components/snackbars/).

**Reply by fdinhSG** [reply]
Thank you for your reply! Adding the VContainer solved my layout issues :)

Do you know why the labels aren't appearing to the numeric VTextField? If I define one as `vuetify.VTextField(label='Min P', v_model=('Pmin', 0), type='number', single_line=True, change=update_bounds)` the label isn't appearing on top of it like when I'm not using `type='number'`:
![image](https://user-images.githubusercontent.com/128691970/233666216-ae181680-2b6d-4a95-a499-7cfb267ef1e8.png)


Regarding the VSnackbar, do I need to `import html` in my script? If I do so, the following error message appears: `AttributeError: module 'html' has no attribute 'Div'`

And thanks again!

**Reply by fdinhSG** [reply]
@jourdain never mind about the label, I had to remove `single_line=True`! I'm still kinda puzzled about `html.Div` though

**Reply by jourdain** [reply]
What about `single_line=True`?

If you want to use the html elements, then yes. (`from trame.widgets import html`)

**Reply by jourdain** [reply]
BTW, I've edited my response to not have any external dependency by just inlining the text content... Just look at the discussion not your email...

**Reply by fdinhSG** [reply]
Thank you! Everything is working properly now :)
